File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/rules/RequireEnvShebangUnderRule.java
package ai.harness.maven.rules;

import ai.harness.maven.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Collection;
import java.util.stream.Stream;

/**
 * Rule that requires executable scripts to use /usr/bin/env shebang.
 */
public enum RequireEnvShebangUnderRule implements HarnessCheckRule {
    INSTANCE;
    private static final String CATEGORY = "requireEnvShebangUnder";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        JsonNode catNode = manifest.get(CATEGORY);
        List<String> directories = HarnessCheckHelper.extractPaths(catNode.get("parameters").get("directories"));
        String expectedPrefix = catNode.get("parameters").get("expectedPrefix").asText();
        String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return directories.stream()
                .flatMap(dir -> {
                    try {
                        Path dirPath = root.resolve(dir);
                        return HarnessCheckHelper.safeFileOrWalk(root, dirPath).stream()
                                .filter(Files::isExecutable)
                                .flatMap(file -> validateShebang(root, file, expectedPrefix, severity).stream());
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .toList();
    }

    private List<Finding> validateShebang(Path root, Path file, String expectedPrefix, String severity) throws MojoExecutionException {
        String text = HarnessCheckHelper.readFile(root, file);
        List<Finding> findings;
        if (text.startsWith("#!") && !text.startsWith(expectedPrefix)) {
            findings = List.of(new Finding(severity, CATEGORY, "executable script should use /usr/bin/env shebang: " + root.relativize(file)));
        } else {
            findings = List.of();
        }
        return findings;
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/rules/RequireEnvShebangUnderRule.kt
package ai.harness.gradle.rules

import ai.harness.gradle.Finding
import ai.harness.gradle.HarnessCheck
import ai.harness.gradle.HarnessPsiResults
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isDirectory
import kotlin.io.path.isExecutable
import kotlin.io.path.readText
import kotlin.io.path.relativeTo

/**
 * Rule that requires executable scripts under certain directories to use /usr/bin/env shebang.
 */
object RequireEnvShebangUnderRule : HarnessCheckRule {
    override fun applies(manifest: JsonObject): Boolean {
        val category = "requireEnvShebangUnder"
        val catObj = manifest[category]?.jsonObject ?: return false
        val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
        return enabled
    }

    override fun validate(
        manifest: JsonObject,
        root: Path,
        psiResults: HarnessPsiResults?,
    ): Collection<Finding> {
        val category = "requireEnvShebangUnder"
        val catObj = manifest[category]?.jsonObject
        val parametersObj = catObj?.get("parameters")?.jsonObject
        val messagesObj = catObj?.get("messages")?.jsonObject
        return if (catObj == null || parametersObj == null || messagesObj == null) {
            emptyList()
        } else {
            val directories = HarnessCheck.stringArrayFrom(parametersObj, "directories")
            val expectedPrefix = HarnessCheck.stringFrom(parametersObj, "expectedPrefix")
            directories
                .filter { dirPath ->
                    (root / dirPath).isDirectory()
                }.flatMap { dirPath ->
                    val dir = root / dirPath
                    val (files, _) = HarnessCheck.walkSafe(root, dir)
                    files
                        .filter { file -> file.isExecutable() }
                        .filter { file ->
                            val firstLine = file.readText().lineSequence().firstOrNull() ?: ""
                            firstLine.startsWith("#!") && !firstLine.startsWith(expectedPrefix)
                        }.map { file ->
                            Finding(
                                HarnessCheck.severityOf(manifest, category),
                                category,
                                HarnessCheck.stringFrom(messagesObj, "default").takeIf { message -> message.isNotEmpty() }
                                    ?: "executable script should use /usr/bin/env shebang: ${file.relativeTo(root)}",
                            )
                        }
                }
        }
    }
}
